File: lib/models/loyalty.dart
class LoyaltyProgram {
  final String userId;
  final int totalCredits;
  final int availableCredits;
  final LoyaltyTier tier;
  final int loginStreak;
  final DateTime? lastLoginDate;
  final double totalSpent;
  final int totalOrders;
  final DateTime joinedAt;

  LoyaltyProgram({
    required this.userId,
    required this.totalCredits,
    required this.availableCredits,
    required this.tier,
    required this.loginStreak,
    this.lastLoginDate,
    required this.totalSpent,
    required this.totalOrders,
    required this.joinedAt,
  });

  // Calculate spending to next tier
  double get spendingToNextTier {
    switch (tier) {
      case LoyaltyTier.bronze:
        return 1000 - totalSpent;
      case LoyaltyTier.silver:
        return 5000 - totalSpent;
      case LoyaltyTier.gold:
        return 10000 - totalSpent;
      case LoyaltyTier.platinum:
        return 0; // Already at max tier
    }
  }

  // Calculate tier progress percentage
  double get tierProgress {
    switch (tier) {
      case LoyaltyTier.bronze:
        return (totalSpent / 1000).clamp(0.0, 1.0);
      case LoyaltyTier.silver:
        return ((totalSpent - 1000) / 4000).clamp(0.0, 1.0);
      case LoyaltyTier.gold:
        return ((totalSpent - 5000) / 5000).clamp(0.0, 1.0);
      case LoyaltyTier.platinum:
        return 1.0;
    }
  }

  // Check if eligible for streak bonus
  bool get isStreakActive {
    if (lastLoginDate == null) return false;
    final now = DateTime.now();
    final difference = now.difference(lastLoginDate!).inDays;
    return difference <= 1;
  }

  factory LoyaltyProgram.fromJson(Map<String, dynamic> json) {
    return LoyaltyProgram(
      userId: json['userId'],
      totalCredits: json['totalCredits'] ?? 0,
      availableCredits: json['availableCredits'] ?? 0,
      tier: LoyaltyTier.values.firstWhere(
        (t) => t.toString() == 'LoyaltyTier.${json['tier']}',
        orElse: () => LoyaltyTier.bronze,
      ),
      loginStreak: json['loginStreak'] ?? 0,
      lastLoginDate: json['lastLoginDate'] != null
          ? DateTime.parse(json['lastLoginDate'])
          : null,
      totalSpent: (json['totalSpent'] ?? 0).toDouble(),
      totalOrders: json['totalOrders'] ?? 0,
      joinedAt: json['joinedAt'] != null
          ? DateTime.parse(json['joinedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'userId': userId,
      'totalCredits': totalCredits,
      'availableCredits': availableCredits,
      'tier': tier.toString().split('.').last,
      'loginStreak': loginStreak,
      'lastLoginDate': lastLoginDate?.toIso8601String(),
      'totalSpent': totalSpent,
      'totalOrders': totalOrders,
      'joinedAt': joinedAt.toIso8601String(),
    };
  }
}

enum LoyaltyTier {
  bronze,
  silver,
  gold,
  platinum,
}

extension LoyaltyTierExtension on LoyaltyTier {
  String get displayName {
    switch (this) {
      case LoyaltyTier.bronze:
        return 'Bronze';
      case LoyaltyTier.silver:
        return 'Silver';
      case LoyaltyTier.gold:
        return 'Gold';
      case LoyaltyTier.platinum:
        return 'Platinum';
    }
  }

  String get icon {
    switch (this) {
      case LoyaltyTier.bronze:
        return '🥉';
      case LoyaltyTier.silver:
        return '🥈';
      case LoyaltyTier.gold:
        return '🥇';
      case LoyaltyTier.platinum:
        return '💎';
    }
  }

  double get creditsMultiplier {
    switch (this) {
      case LoyaltyTier.bronze:
        return 1.0;
      case LoyaltyTier.silver:
        return 1.5;
      case LoyaltyTier.gold:
        return 2.0;
      case LoyaltyTier.platinum:
        return 2.5;
    }
  }

  double get spendingRequired {
    switch (this) {
      case LoyaltyTier.bronze:
        return 0;
      case LoyaltyTier.silver:
        return 1000;
      case LoyaltyTier.gold:
        return 5000;
      case LoyaltyTier.platinum:
        return 10000;
    }
  }

  List<String> get benefits {
    switch (this) {
      case LoyaltyTier.bronze:
        return [
          'Earn 1 credit per \$1 spent',
          'Login streak bonuses',
          'Early access to sales',
        ];
      case LoyaltyTier.silver:
        return [
          'Earn 1.5 credits per \$1 spent',
          'Free shipping on orders over \$50',
          'Exclusive member discounts',
          'Enhanced login bonuses',
        ];
      case LoyaltyTier.gold:
        return [
          'Earn 2 credits per \$1 spent',
          'Free shipping on all orders',
          'Priority customer service',
          'Exclusive gold member sales',
          'Double login bonuses',
        ];
      case LoyaltyTier.platinum:
        return [
          'Earn 2.5 credits per \$1 spent',
          'Free express shipping',
          'VIP customer service',
          'Exclusive platinum previews',
          'Personal shopping assistant',
          'Triple login bonuses',
        ];
    }
  }
}

class CreditTransaction {
  final String id;
  final String userId;
  final TransactionType type;
  final int credits;
  final String description;
  final DateTime createdAt;
  final String? orderId;
  final String? voucherId;

  CreditTransaction({
    required this.id,
    required this.userId,
    required this.type,
    required this.credits,
    required this.description,
    required this.createdAt,
    this.orderId,
    this.voucherId,
  });

  bool get isPositive => credits > 0;

  factory CreditTransaction.fromJson(Map<String, dynamic> json) {
    return CreditTransaction(
      id: json['id'] ?? '',
      userId: json['userId'] ?? '',
      type: TransactionType.values.firstWhere(
        (t) => t.toString() == 'TransactionType.${json['type']}',
        orElse: () => TransactionType.earned,
      ),
      credits: json['credits'] ?? 0,
      description: json['description'] ?? '',
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      orderId: json['orderId'],
      voucherId: json['voucherId'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'type': type.toString().split('.').last,
      'credits': credits,
      'description': description,
      'createdAt': createdAt.toIso8601String(),
      if (orderId != null) 'orderId': orderId,
      if (voucherId != null) 'voucherId': voucherId,
    };
  }
}

enum TransactionType {
  earned,
  redeemed,
  loginBonus,
  streakBonus,
  welcomeBonus,
  expired,
  refund,
}

extension TransactionTypeExtension on TransactionType {
  String get displayName {
    switch (this) {
      case TransactionType.earned:
        return 'Earned';
      case TransactionType.redeemed:
        return 'Redeemed';
      case TransactionType.loginBonus:
        return 'Login Bonus';
      case TransactionType.streakBonus:
        return 'Streak Bonus';
      case TransactionType.welcomeBonus:
        return 'Welcome Bonus';
      case TransactionType.expired:
        return 'Expired';
      case TransactionType.refund:
        return 'Refund';
    }
  }
}

class RedemptionOption {
  final String id;
  final String name;
  final String description;
  final int creditsRequired;
  final double discountValue;
  final RedemptionType type;

  RedemptionOption({
    required this.id,
    required this.name,
    required this.description,
    required this.creditsRequired,
    required this.discountValue,
    required this.type,
  });

  factory RedemptionOption.fromJson(Map<String, dynamic> json) {
    return RedemptionOption(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      creditsRequired: json['creditsRequired'] ?? 0,
      discountValue: (json['discountValue'] ?? 0).toDouble(),
      type: RedemptionType.values.firstWhere(
        (t) => t.toString() == 'RedemptionType.${json['type']}',
        orElse: () => RedemptionType.discount,
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'creditsRequired': creditsRequired,
      'discountValue': discountValue,
      'type': type.toString().split('.').last,
    };
  }
}

enum RedemptionType {
  discount,
  voucher,
  freeShipping,
}

extension RedemptionTypeExtension on RedemptionType {
  String get displayName {
    switch (this) {
      case RedemptionType.discount:
        return 'Discount';
      case RedemptionType.voucher:
        return 'Voucher';
      case RedemptionType.freeShipping:
        return 'Free Shipping';
    }
  }
}

class TierInfo {
  final String name;
  final String icon;
  final double spendingRequired;
  final double creditsMultiplier;
  final List<String> benefits;

  TierInfo({
    required this.name,
    required this.icon,
    required this.spendingRequired,
    required this.creditsMultiplier,
    required this.benefits,
  });

  factory TierInfo.fromJson(Map<String, dynamic> json) {
    return TierInfo(
      name: json['name'] ?? '',
      icon: json['icon'] ?? '',
      spendingRequired: (json['spendingRequired'] ?? 0).toDouble(),
      creditsMultiplier: (json['creditsMultiplier'] ?? 1.0).toDouble(),
      benefits: (json['benefits'] as List<dynamic>?)
              ?.map((b) => b.toString())
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'icon': icon,
      'spendingRequired': spendingRequired,
      'creditsMultiplier': creditsMultiplier,
      'benefits': benefits,
    };
  }
}

File: backend/internal/models/loyalty.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// LoyaltyProgram represents a user's loyalty program status
type LoyaltyProgram struct {
	ID               primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	UserID           primitive.ObjectID `json:"userId" bson:"userId"`
	TotalCredits     int                `json:"totalCredits" bson:"totalCredits"`         // Lifetime credits earned
	AvailableCredits int                `json:"availableCredits" bson:"availableCredits"` // Current balance
	Tier             LoyaltyTier        `json:"tier" bson:"tier"`
	LoginStreak      int                `json:"loginStreak" bson:"loginStreak"`
	LastLoginDate    *time.Time         `json:"lastLoginDate" bson:"lastLoginDate"`
	TotalSpent       float64            `json:"totalSpent" bson:"totalSpent"`
	TotalOrders      int                `json:"totalOrders" bson:"totalOrders"`
	JoinedAt         time.Time          `json:"joinedAt" bson:"joinedAt"`
	UpdatedAt        time.Time          `json:"updatedAt" bson:"updatedAt"`
}

// LoyaltyTier represents different tiers in the loyalty program
type LoyaltyTier string

const (
	TierBronze   LoyaltyTier = "bronze"
	TierSilver   LoyaltyTier = "silver"
	TierGold     LoyaltyTier = "gold"
	TierPlatinum LoyaltyTier = "platinum"
)

// TierInfo provides tier-specific information
type TierInfo struct {
	Name             string   `json:"name"`
	Icon             string   `json:"icon"`
	SpendingRequired float64  `json:"spendingRequired"` // Total spending needed for tier
	CreditsMultiplier float64 `json:"creditsMultiplier"`
	Benefits         []string `json:"benefits"`
}

// GetTierInfo returns information for a given tier
func (t LoyaltyTier) GetTierInfo() TierInfo {
	switch t {
	case TierBronze:
		return TierInfo{
			Name:              "Bronze",
			Icon:              "🥉",
			SpendingRequired:  0,
			CreditsMultiplier: 1.0,
			Benefits: []string{
				"Earn 1 credit per $1 spent",
				"Login streak bonuses",
				"Early access to sales",
			},
		}
	case TierSilver:
		return TierInfo{
			Name:              "Silver",
			Icon:              "🥈",
			SpendingRequired:  1000,
			CreditsMultiplier: 1.5,
			Benefits: []string{
				"Earn 1.5 credits per $1 spent",
				"Free shipping on orders over $50",
				"Exclusive member discounts",
				"Enhanced login bonuses",
			},
		}
	case TierGold:
		return TierInfo{
			Name:              "Gold",
			Icon:              "🥇",
			SpendingRequired:  5000,
			CreditsMultiplier: 2.0,
			Benefits: []string{
				"Earn 2 credits per $1 spent",
				"Free shipping on all orders",
				"Priority customer service",
				"Exclusive gold member sales",
				"Double login bonuses",
			},
		}
	case TierPlatinum:
		return TierInfo{
			Name:              "Platinum",
			Icon:              "💎",
			SpendingRequired:  10000,
			CreditsMultiplier: 2.5,
			Benefits: []string{
				"Earn 2.5 credits per $1 spent",
				"Free express shipping",
				"VIP customer service",
				"Exclusive platinum previews",
				"Personal shopping assistant",
				"Triple login bonuses",
			},
		}
	default:
		return TierInfo{}
	}
}

// CalculateNextTier returns the next tier and spending needed
func (lp *LoyaltyProgram) CalculateNextTier() (LoyaltyTier, float64) {
	switch lp.Tier {
	case TierBronze:
		return TierSilver, 1000 - lp.TotalSpent
	case TierSilver:
		return TierGold, 5000 - lp.TotalSpent
	case TierGold:
		return TierPlatinum, 10000 - lp.TotalSpent
	default:
		return TierPlatinum, 0 // Already at max tier
	}
}

// CalculateTierProgress returns progress percentage for current tier
func (lp *LoyaltyProgram) CalculateTierProgress() float64 {
	switch lp.Tier {
	case TierBronze:
		return (lp.TotalSpent / 1000.0) * 100
	case TierSilver:
		return ((lp.TotalSpent - 1000) / 4000.0) * 100
	case TierGold:
		return ((lp.TotalSpent - 5000) / 5000.0) * 100
	case TierPlatinum:
		return 100.0
	default:
		return 0.0
	}
}

// UpdateTierBasedOnSpending updates the tier based on total spending
func (lp *LoyaltyProgram) UpdateTierBasedOnSpending() {
	if lp.TotalSpent >= 10000 {
		lp.Tier = TierPlatinum
	} else if lp.TotalSpent >= 5000 {
		lp.Tier = TierGold
	} else if lp.TotalSpent >= 1000 {
		lp.Tier = TierSilver
	} else {
		lp.Tier = TierBronze
	}
}

// CreditTransaction represents a single credit transaction
type CreditTransaction struct {
	ID          primitive.ObjectID   `json:"id" bson:"_id,omitempty"`
	UserID      primitive.ObjectID   `json:"userId" bson:"userId"`
	Type        TransactionType      `json:"type" bson:"type"`
	Credits     int                  `json:"credits" bson:"credits"`
	Description string               `json:"description" bson:"description"`
	OrderID     *primitive.ObjectID  `json:"orderId,omitempty" bson:"orderId,omitempty"`
	VoucherID   *primitive.ObjectID  `json:"voucherId,omitempty" bson:"voucherId,omitempty"`
	CreatedAt   time.Time            `json:"createdAt" bson:"createdAt"`
}

// TransactionType represents different types of credit transactions
type TransactionType string

const (
	TransactionEarned       TransactionType = "earned"        // Earned from purchase
	TransactionRedeemed     TransactionType = "redeemed"      // Redeemed for discount/voucher
	TransactionLoginBonus   TransactionType = "login_bonus"   // Daily login bonus
	TransactionStreakBonus  TransactionType = "streak_bonus"  // Login streak milestone bonus
	TransactionWelcomeBonus TransactionType = "welcome_bonus" // Welcome bonus
	TransactionExpired      TransactionType = "expired"       // Credits expired
	TransactionRefund       TransactionType = "refund"        // Refunded from cancelled order
)

// PointTransaction is an alias for backward compatibility
type PointTransaction = CreditTransaction

// VoucherType represents different types of vouchers
type VoucherType string

const (
	VoucherPercentage   VoucherType = "percentage"
	VoucherFixed        VoucherType = "fixed"
	VoucherFreeShipping VoucherType = "freeShipping"
	VoucherGift         VoucherType = "gift"
)

// LoyaltyConfig represents configuration for the loyalty program
type LoyaltyConfig struct {
	ID                   primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	BaseCreditsPerDollar float64            `json:"baseCreditsPerDollar" bson:"baseCreditsPerDollar"`
	DailyLoginBonus      int                `json:"dailyLoginBonus" bson:"dailyLoginBonus"`
	StreakBonusCredits   int                `json:"streakBonusCredits" bson:"streakBonusCredits"`
	StreakBonusDays      int                `json:"streakBonusDays" bson:"streakBonusDays"` // Award bonus every X days
	WelcomeBonus         int                `json:"welcomeBonus" bson:"welcomeBonus"`
	CreditsToMoneyRatio  float64            `json:"creditsToMoneyRatio" bson:"creditsToMoneyRatio"` // e.g., 100 credits = $1
	UpdatedAt            time.Time          `json:"updatedAt" bson:"updatedAt"`
}

// Default loyalty configuration
func DefaultLoyaltyConfig() *LoyaltyConfig {
	return &LoyaltyConfig{
		BaseCreditsPerDollar: 1.0,   // 1 credit per dollar spent
		DailyLoginBonus:      10,    // 10 credits for daily login
		StreakBonusCredits:   50,    // 50 bonus credits for streak milestone
		StreakBonusDays:      7,     // Bonus every 7 days of streak
		WelcomeBonus:         100,   // 100 credits on signup
		CreditsToMoneyRatio:  100.0, // 100 credits = $1 discount
		UpdatedAt:            time.Now(),
	}
}

// RedemptionOption represents options for redeeming credits
type RedemptionOption struct {
	ID          string  `json:"id"`
	Name        string  `json:"name"`
	Description string  `json:"description"`
	CreditsRequired int `json:"creditsRequired"`
	DiscountValue   float64 `json:"discountValue"`
	Type        RedemptionType `json:"type"`
}

// RedemptionType represents the type of redemption
type RedemptionType string

const (
	RedemptionDiscount      RedemptionType = "discount"        // Direct discount on order
	RedemptionVoucher       RedemptionType = "voucher"         // Generate voucher code
	RedemptionFreeShipping  RedemptionType = "free_shipping"   // Free shipping voucher
)

// GetRedemptionOptions returns available redemption options
func GetRedemptionOptions() []RedemptionOption {
	return []RedemptionOption{
		{
			ID:              "discount_5",
			Name:            "$5 Discount",
			Description:     "Get $5 off your order",
			CreditsRequired: 500,
			DiscountValue:   5.0,
			Type:            RedemptionDiscount,
		},
		{
			ID:              "discount_10",
			Name:            "$10 Discount",
			Description:     "Get $10 off your order",
			CreditsRequired: 1000,
			DiscountValue:   10.0,
			Type:            RedemptionDiscount,
		},
		{
			ID:              "discount_25",
			Name:            "$25 Discount",
			Description:     "Get $25 off your order",
			CreditsRequired: 2500,
			DiscountValue:   25.0,
			Type:            RedemptionDiscount,
		},
		{
			ID:              "free_shipping",
			Name:            "Free Shipping",
			Description:     "Get free shipping on your next order",
			CreditsRequired: 300,
			DiscountValue:   0,
			Type:            RedemptionFreeShipping,
		},
	}
}

// LoyaltyStatistics represents statistics for the loyalty program
type LoyaltyStatistics struct {
	TotalMembers          int64            `json:"totalMembers"`
	ActiveMembers         int64            `json:"activeMembers"`
	TierDistribution      map[string]int64 `json:"tierDistribution"`
	TotalCreditsIssued    int64            `json:"totalCreditsIssued"`
	TotalCreditsRedeemed  int64            `json:"totalCreditsRedeemed"`
	AverageCreditsPerUser float64          `json:"averageCreditsPerUser"`
}

// TopLoyaltyMember represents a top loyalty program member
type TopLoyaltyMember struct {
	UserID           primitive.ObjectID `json:"userId"`
	TotalCredits     int                `json:"totalCredits"`
	AvailableCredits int                `json:"availableCredits"`
	Tier             LoyaltyTier        `json:"tier"`
	TotalSpent       float64            `json:"totalSpent"`
}